docs/public-async-nio-apis.md

```diff
@@ -0,0 +1,1162 @@
+    /// elements of the asynchronous sequence.
+    @inlinable public func makeAsyncIterator() -> NIOCore.NIOAsyncChannelInboundStream<Inbound>.AsyncIterator
+}
+
+/// A ``NIOAsyncChannelOutboundWriter`` is used to write and flush new outbound messages in a channel.
+///
+/// The writer acts as a bridge between the Concurrency and NIO world. It allows to write and flush messages into the
+/// underlying ``Channel``. Furthermore, it respects back-pressure of the channel by suspending the calls to write until
+/// the channel becomes writable again.
+@available(macOS 10.15, iOS 13, tvOS 13, watchOS 6, *)
+public struct NIOAsyncChannelOutboundWriter<OutboundOut> : Sendable where OutboundOut : Sendable {
+    /// An `AsyncSequence` backing a ``NIOAsyncChannelOutboundWriter`` for testing purposes.
+    public struct TestSink : AsyncSequence {
+        /// The type of element produced by this asynchronous sequence.
+        public typealias Element = OutboundOut
+
+        /// Creates the asynchronous iterator that produces elements of this
+        /// asynchronous sequence.
+        ///
+        /// - Returns: An instance of the `AsyncIterator` type used to produce
+        /// elements of the asynchronous sequence.
+        public func makeAsyncIterator() -> NIOCore.NIOAsyncChannelOutboundWriter<OutboundOut>.TestSink.AsyncIterator
+
+        /// The type of asynchronous iterator that produces elements of this
+        /// asynchronous sequence.
+        public struct AsyncIterator : AsyncIteratorProtocol {
+            /// Asynchronously advances to the next element and returns it, or ends the
+            /// sequence if there is no next element.
+            /// 
+            /// - Returns: The next element, if it exists, or `nil` to signal the end of
+            ///   the sequence.
+            public mutating func next() async -> NIOCore.NIOAsyncChannelOutboundWriter<OutboundOut>.TestSink.Element?
+        }
+    }
+
+    /// Creates a new ``NIOAsyncChannelOutboundWriter`` backed by a ``NIOAsyncChannelOutboundWriter/TestSink``.
+    /// This is mostly useful for testing purposes where one wants to observe the written data.
+    @inlinable public static func makeTestingWriter() -> (NIOCore.NIOAsyncChannelOutboundWriter<OutboundOut>, NIOCore.NIOAsyncChannelOutboundWriter<OutboundOut>.TestSink)
+
+    /// Send a write into the ``ChannelPipeline`` and flush it right away.
+    ///
+    /// This method suspends if the underlying channel is not writable and will resume once the it becomes writable again.
+    @inlinable public func write(_ data: OutboundOut) async throws
+
+    /// Send a sequence of writes into the ``ChannelPipeline`` and flush them right away.
+    ///
+    /// This method suspends if the underlying channel is not writable and will resume once the it becomes writable again.
+    @inlinable public func write<Writes>(contentsOf sequence: Writes) async throws where OutboundOut == Writes.Element, Writes : Sequence
```

Comment:
But this case is super important for performance sensitive cases otherwise we lose the batch writing API. FWIW, Rust has the same on their [Writer](https://doc.rust-lang.org/std/io/trait.Write.html) training

Reply:
I'm absolutely not saying we shouldn't have this – I agree it's really important. 

I'm just noting that it's a shame we have to special case this, especially as on the read side we have to suspend for every message.